--- a/target/linux/sunxi/patches-4.14/203-dts-add-usb2-usb3-soc_codec-opi_zero.patch	1970-01-01 03:00:00.000000000 +0300
+++ b/target/linux/sunxi/patches-4.14/203-dts-add-usb2-usb3-soc_codec-opi_zero.patch	2019-04-18 18:49:47.000000000 +0300
@@ -0,0 +1,42 @@
+--- a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
++++ b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
+@@ -103,6 +103,14 @@
+ &ehci1 {
+ 	status = "okay";
+ };
++
++&ehci2 {
++	status = "okay";
++};
++
++&ehci3 {
++	status = "okay";
++};
+ 
+ &emac {
+ 	phy-handle = <&int_mii_phy>;
+@@ -150,10 +150,23 @@
+ &ohci1 {
+ 	status = "okay";
+ };
++
++&ohci2 {
++	status = "okay";
++};
++
++&ohci3 {
++	status = "okay";
++};
++
++&codec {
++	allwinner,audio-routing = "Line Out", "LINEOUT", "MIC1", "Mic", "Mic", "MBIAS";
++	status = "okay";
++};
+ 
+ &spi0 {
+ 	/* Disable SPI NOR by default: it optional on Orange Pi Zero boards */
+-	status = "disabled";
++	status = "okay";
+ 
+ 	flash@0 {
+ 		#address-cells = <1>;
--- a/target/linux/sunxi/patches-4.14/204-dts-wifi-add-xradio.patch.patch	1970-01-01 03:00:00.000000000 +0300
+++ b/target/linux/sunxi/patches-4.14/204-dts-wifi-add-xradio.patch.patch	2019-04-12 14:17:28.000000000 +0300
@@ -0,0 +1,47 @@
+diff --git a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
+index 454408b..df356b5 100644
+--- a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
++++ b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
+@@ -49,6 +49,7 @@
+ 
+ #include <dt-bindings/gpio/gpio.h>
+ #include <dt-bindings/input/input.h>
++#include <dt-bindings/pinctrl/sun4i-a10.h>
+ 
+ / {
+ 	model = "Xunlong Orange Pi Zero";
+@@ -88,6 +88,23 @@
+ 		enable-active-high;
+ 		gpio = <&pio 0 20 GPIO_ACTIVE_HIGH>;
+ 	};
++
++	reg_vdd_cpux: vdd-cpux-regulator {
++		compatible = "regulator-gpio";
++		regulator-name = "vdd-cpux";
++		regulator-type = "voltage";
++		regulator-boot-on;
++		regulator-always-on;
++		regulator-min-microvolt = <1100000>;
++		regulator-max-microvolt = <1300000>;
++		regulator-ramp-delay = <50>; /* 4ms */
++
++		gpios = <&r_pio 0 6 GPIO_ACTIVE_HIGH>; /* PL6 */
++		enable-active-high;
++		gpios-states = <1>;
++		states = <1100000 0
++			  1300000 1>;
++	};
+ 
+ 	wifi_pwrseq: wifi_pwrseq {
+ 		compatible = "mmc-pwrseq-simple";
+@@ -144,6 +144,10 @@
+ 	 */
+ 	xr819: sdio_wifi@1 {
+ 		reg = <1>;
++		compatible = "xradio,xr819";
++		interrupt-parent = <&pio>;
++		interrupts = <6 10 IRQ_TYPE_EDGE_RISING>;
++		interrupt-names = "host-wake";
+ 	};
+ };
+ 
--- a/target/linux/sunxi/image/cortex-a7.mk	2019-04-18 22:36:29.077431485 +0300
+++ b/target/linux/sunxi/image/cortex-a7.mk	2019-04-18 19:34:54.743409400 +0300
@@ -110,7 +110,12 @@
 
 define Device/sun8i-h2-plus-orangepi-zero
   DEVICE_TITLE:=Xunlong Orange Pi Zero
-  DEVICE_PACKAGES:=kmod-rtc-sunxi
+  DEVICE_PACKAGES:=kmod-rtc-sunxi \
+		kmod-xradio \
+		kmod-sound-core \
+		kmod-sound-soc-sun4i-codec \
+		kmod-sound-soc-sun8i-codec \
+		kmod-sound-soc-sun8i-codec-analog
   SUPPORTED_DEVICES:=xunlong,orangepi-zero
   SUNXI_DTS:=sun8i-h2-plus-orangepi-zero
 endef
--- a/target/linux/sunxi/config-4.14	2019-04-18 22:36:29.077431485 +0300
+++ b/target/linux/sunxi/config-4.14	2019-04-18 19:35:39.455631116 +0300
@@ -503,8 +503,6 @@
 CONFIG_SND_SOC_I2C_AND_SPI=y
 # CONFIG_SND_SUN4I_I2S is not set
 # CONFIG_SND_SUN4I_SPDIF is not set
-# CONFIG_SND_SUN8I_CODEC is not set
-# CONFIG_SND_SUN8I_CODEC_ANALOG is not set
 CONFIG_SOUND=y
 CONFIG_SOUND_OSS_CORE=y
 # CONFIG_SOUND_OSS_CORE_PRECLAIM is not set
--- a/target/linux/sunxi/modules.mk	2019-04-18 22:36:29.077431485 +0300
+++ b/target/linux/sunxi/modules.mk	2019-04-18 19:36:24.599854974 +0300
@@ -69,17 +69,49 @@
 $(eval $(call KernelPackage,sun4i-emac))
 
 
-define KernelPackage/sound-soc-sunxi
-  TITLE:=AllWinner built-in SoC sound support
+define KernelPackage/sound-soc-sun4i-codec
+  TITLE:=AllWinner built-in SoC sound support sun4i-codec
   KCONFIG:=CONFIG_SND_SUN4I_CODEC
   FILES:=$(LINUX_DIR)/sound/soc/sunxi/sun4i-codec.ko
-  AUTOLOAD:=$(call AutoLoad,65,sun4i-codec)
+  AUTOLOAD:=$(call AutoLoad,63,sun4i-codec)
   DEPENDS:=@TARGET_sunxi +kmod-sound-soc-core
   $(call AddDepends/sound)
 endef
 
-define KernelPackage/sound-soc-sunxi/description
-  Kernel support for AllWinner built-in SoC audio
+define KernelPackage/sound-soc-sun4i-codec/description
+  Kernel support for AllWinner built-in SoC audio sun4i-codec
 endef
 
-$(eval $(call KernelPackage,sound-soc-sunxi))
+$(eval $(call KernelPackage,sound-soc-sun4i-codec))
+
+
+define KernelPackage/sound-soc-sun8i-codec
+  TITLE:=AllWinner SoC sound sun8i-codec
+  KCONFIG:=CONFIG_SND_SUN8I_CODEC
+  FILES:=$(LINUX_DIR)/sound/soc/sunxi/sun8i-codec.ko
+  AUTOLOAD:=$(call AutoLoad,65,sun8i-codec)
+  DEPENDS:=@TARGET_sunxi +kmod-sound-soc-core
+  $(call AddDepends/sound)
+endef
+
+define KernelPackage/sound-soc-sun8i-codec/description
+  Kernel support for AllWinner built-in SoC audio sun8i-codec
+endef
+
+$(eval $(call KernelPackage,sound-soc-sun8i-codec))
+
+
+define KernelPackage/sound-soc-sun8i-codec-analog
+  TITLE:=AllWinner SoC sound sun8i-codec-analog
+  KCONFIG:=CONFIG_SND_SUN8I_CODEC_ANALOG
+  FILES:=$(LINUX_DIR)/sound/soc/sunxi/sun8i-codec-analog.ko
+  AUTOLOAD:=$(call AutoLoad,67,sun8i-codec-analog)
+  DEPENDS:=@TARGET_sunxi +kmod-sound-soc-core
+  $(call AddDepends/sound)
+endef
+
+define KernelPackage/sound-soc-sun8i-codec-analog/description
+  Kernel support for AllWinner built-in SoC audio sun8i-codec-analog
+endef
+
+$(eval $(call KernelPackage,sound-soc-sun8i-codec-analog))
--- a/package/kernel/mac80211/Makefile	2019-04-18 22:36:26.473418573 +0300
+++ b/package/kernel/mac80211/Makefile	2019-04-18 23:04:18.349708959 +0300
@@ -10,7 +10,7 @@
 
 PKG_NAME:=mac80211
 
-PKG_VERSION:=4.19.32-1
+PKG_VERSION:=4.19.32-2
 PKG_RELEASE:=2
 PKG_SOURCE_URL:=@KERNEL/linux/kernel/projects/backports/stable/v4.19.32/
 PKG_HASH:=838db1565b54fe4bd1e769c54f30c65c9ea2fb5e99a0cddb7910561794ae317a
@@ -31,6 +31,7 @@
 	p54-common p54-pci p54-spi p54-usb \
 	rsi91x rsi91x-usb rsi91x-sdio\
 	wlcore wl12xx wl18xx \
+	xradio \
 	zd1211rw
 
 PKG_CONFIG_DEPENDS:= \
@@ -77,6 +78,7 @@
 	WLAN_VENDOR_RSI \
 	WLAN_VENDOR_ST \
 	WLAN_VENDOR_TI \
+	WLAN_VENDOR_XRADIO \
 	WLAN_VENDOR_ZYDAS \
 
 config-$(call config_package,cfg80211) += CFG80211
@@ -386,6 +388,14 @@
   AUTOLOAD:=$(call AutoProbe,zd1211rw)
 endef
 
+define KernelPackage/xradio
+  $(call KernelPackage/mac80211/Default)
+  TITLE:=Xradio xr819 test wifi-drivers support
+  DEPENDS+=+kmod-mac80211
+  FILES:=$(PKG_BUILD_DIR)/drivers/net/wireless/xradio/xradio_wlan.ko
+  AUTOLOAD:=$(call AutoProbe,xradio_wlan)
+endef
+
 ifdef CONFIG_PACKAGE_MAC80211_DEBUGFS
   config-y += \
 	CFG80211_DEBUGFS \
@@ -421,6 +431,7 @@
 config-$(call config_package,wl12xx) += WL12XX
 config-$(call config_package,wl18xx) += WL18XX
 config-y += WL_TI WILINK_PLATFORM_DATA
+config-$(call config_package,xradio) += XRADIO
 config-$(call config_package,zd1211rw) += ZD1211RW
 config-$(call config_package,rsi91x) += RSI_91X
 config-$(call config_package,rsi91x-usb) += RSI_USB
@@ -565,6 +576,17 @@
 	$(INSTALL_DATA) $(PKG_BUILD_DIR)/$(ZD1211FW_NAME)/zd1211* $(1)/lib/firmware/zd1211
 endef
 
+define KernelPackage/xradio/install
+#	$(INSTALL_DIR) $(1)/lib/modules/$(LINUX_VERSION)
+#	$(INSTALL_BIN) \
+#		$(PKG_BUILD_DIR)/drivers/net/wireless/xradio/xradio_wlan.ko \
+#			$(1)/lib/modules/$(LINUX_VERSION)/
+#
+	$(INSTALL_DIR) $(1)/lib/firmware/xr819
+	$(CP) $(PKG_BUILD_DIR)/drivers/net/wireless/xradio/xr819/*_xr819.bin \
+		$(1)/lib/firmware/xr819/
+endef
+
 $(eval $(foreach drv,$(PKG_DRIVERS),$(call KernelPackage,$(drv))))
 $(eval $(call KernelPackage,cfg80211))
 $(eval $(call KernelPackage,mac80211))
