--- a/target/linux/sunxi/image/cortex-a7.mk	2019-01-30 14:34:49.000000000 +0200
+++ b/target/linux/sunxi/image/cortex-a7.mk	2019-04-20 16:15:23.904581930 +0300
@@ -108,6 +108,20 @@
 TARGET_DEVICES += sun6i-a31-m9
 
 
+define Device/sun8i-h2-plus-orangepi-zero
+  DEVICE_TITLE:=Xunlong Orange Pi Zero
+  DEVICE_PACKAGES:=kmod-rtc-sunxi \
+		kmod-sound-core \
+		kmod-sound-soc-sun4i-codec \
+		kmod-sound-soc-sun8i-codec \
+		kmod-sound-soc-sun8i-codec-analog
+  SUPPORTED_DEVICES:=xunlong,orangepi-zero
+  SUNXI_DTS:=sun8i-h2-plus-orangepi-zero
+endef
+
+TARGET_DEVICES += sun8i-h2-plus-orangepi-zero
+
+
 define Device/sun8i-h2-plus-orangepi-r1
   DEVICE_TITLE:=Xunlong Orange Pi R1
   DEVICE_PACKAGES:=kmod-rtc-sunxi kmod-usb-net kmod-usb-net-rtl8152
