--- a/package/boot/uboot-sunxi/Makefile	2019-01-30 14:34:49.000000000 +0200
+++ b/package/boot/uboot-sunxi/Makefile	2019-04-20 16:15:23.888581851 +0300
@@ -150,6 +150,12 @@
   BUILD_DEVICES:=sun8i-h2-plus-orangepi-r1
 endef
 
+define U-Boot/orangepi_zero
+  BUILD_SUBTARGET:=cortexa7
+  NAME:=Orange Pi Zero (H2+)
+  BUILD_DEVICES:=sun8i-h2-plus-orangepi-zero
+endef
+
 define U-Boot/orangepi_pc
   BUILD_SUBTARGET:=cortexa7
   NAME:=Orange Pi PC (H3)
@@ -230,6 +236,7 @@
 	nanopi_neo \
 	nanopi_neo_plus2 \
 	orangepi_r1 \
+	orangepi_zero \
 	orangepi_pc \
 	orangepi_plus \
 	orangepi_2 \
