--- a/target/linux/sunxi/patches-4.14/203-dts-add-usb2-usb3-soc_codec-opi_zero.patch	1970-01-01 03:00:00.000000000 +0300
+++ b/target/linux/sunxi/patches-4.14/203-dts-add-usb2-usb3-soc_codec-opi_zero.patch	2019-04-20 16:15:23.896581891 +0300
@@ -0,0 +1,42 @@
+--- a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
++++ b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
+@@ -103,6 +160,14 @@
+ &ehci1 {
+ 	status = "okay";
+ };
++
++&ehci2 {
++	status = "okay";
++};
++
++&ehci3 {
++	status = "okay";
++};
+ 
+ &emac {
+ 	phy-handle = <&int_mii_phy>;
+@@ -150,10 +190,23 @@
+ &ohci1 {
+ 	status = "okay";
+ };
++
++&ohci2 {
++	status = "okay";
++};
++
++&ohci3 {
++	status = "okay";
++};
++
++&codec {
++	allwinner,audio-routing = "Line Out", "LINEOUT", "MIC1", "Mic", "Mic", "MBIAS";
++	status = "okay";
++};
+ 
+ &spi0 {
+ 	/* Disable SPI NOR by default: it optional on Orange Pi Zero boards */
+-	status = "disabled";
++	status = "okay";
+ 
+ 	flash@0 {
+ 		#address-cells = <1>;
--- a/target/linux/sunxi/config-4.14	2019-01-30 14:34:49.000000000 +0200
+++ b/target/linux/sunxi/config-4.14	2019-04-20 16:15:23.924582030 +0300
@@ -495,8 +495,6 @@
 CONFIG_SND_SOC_I2C_AND_SPI=y
 # CONFIG_SND_SUN4I_I2S is not set
 # CONFIG_SND_SUN4I_SPDIF is not set
-# CONFIG_SND_SUN8I_CODEC is not set
-# CONFIG_SND_SUN8I_CODEC_ANALOG is not set
 CONFIG_SOUND=y
 CONFIG_SOUND_OSS_CORE=y
 # CONFIG_SOUND_OSS_CORE_PRECLAIM is not set
--- a/target/linux/sunxi/image/cortex-a7.mk	2019-01-30 14:34:49.000000000 +0200
+++ b/target/linux/sunxi/image/cortex-a7.mk	2019-04-20 16:15:23.904581930 +0300
@@ -108,6 +108,20 @@
 TARGET_DEVICES += sun6i-a31-m9
 
 
+define Device/sun8i-h2-plus-orangepi-zero
+  DEVICE_TITLE:=Xunlong Orange Pi Zero
+  DEVICE_PACKAGES:=kmod-rtc-sunxi \
+		kmod-sound-core \
+		kmod-sound-soc-sun4i-codec \
+		kmod-sound-soc-sun8i-codec \
+		kmod-sound-soc-sun8i-codec-analog
+  SUPPORTED_DEVICES:=xunlong,orangepi-zero
+  SUNXI_DTS:=sun8i-h2-plus-orangepi-zero
+endef
+
+TARGET_DEVICES += sun8i-h2-plus-orangepi-zero
+
+
 define Device/sun8i-h2-plus-orangepi-r1
   DEVICE_TITLE:=Xunlong Orange Pi R1
   DEVICE_PACKAGES:=kmod-rtc-sunxi kmod-usb-net kmod-usb-net-rtl8152
--- a/target/linux/sunxi/modules.mk	2019-01-30 14:34:49.000000000 +0200
+++ b/target/linux/sunxi/modules.mk	2019-04-20 16:15:23.928582049 +0300
@@ -69,17 +69,49 @@
 $(eval $(call KernelPackage,sun4i-emac))
 
 
-define KernelPackage/sound-soc-sunxi
-  TITLE:=AllWinner built-in SoC sound support
+define KernelPackage/sound-soc-sun4i-codec
+  TITLE:=AllWinner built-in SoC sound support sun4i-codec
   KCONFIG:=CONFIG_SND_SUN4I_CODEC
   FILES:=$(LINUX_DIR)/sound/soc/sunxi/sun4i-codec.ko
-  AUTOLOAD:=$(call AutoLoad,65,sun4i-codec)
+  AUTOLOAD:=$(call AutoLoad,63,sun4i-codec)
   DEPENDS:=@TARGET_sunxi +kmod-sound-soc-core
   $(call AddDepends/sound)
 endef
 
-define KernelPackage/sound-soc-sunxi/description
-  Kernel support for AllWinner built-in SoC audio
+define KernelPackage/sound-soc-sun4i-codec/description
+  Kernel support for AllWinner built-in SoC audio sun4i-codec
 endef
 
-$(eval $(call KernelPackage,sound-soc-sunxi))
+$(eval $(call KernelPackage,sound-soc-sun4i-codec))
+
+
+define KernelPackage/sound-soc-sun8i-codec
+  TITLE:=AllWinner SoC sound sun8i-codec
+  KCONFIG:=CONFIG_SND_SUN8I_CODEC
+  FILES:=$(LINUX_DIR)/sound/soc/sunxi/sun8i-codec.ko
+  AUTOLOAD:=$(call AutoLoad,65,sun8i-codec)
+  DEPENDS:=@TARGET_sunxi +kmod-sound-soc-core
+  $(call AddDepends/sound)
+endef
+
+define KernelPackage/sound-soc-sun8i-codec/description
+  Kernel support for AllWinner built-in SoC audio sun8i-codec
+endef
+
+$(eval $(call KernelPackage,sound-soc-sun8i-codec))
+
+
+define KernelPackage/sound-soc-sun8i-codec-analog
+  TITLE:=AllWinner SoC sound sun8i-codec-analog
+  KCONFIG:=CONFIG_SND_SUN8I_CODEC_ANALOG
+  FILES:=$(LINUX_DIR)/sound/soc/sunxi/sun8i-codec-analog.ko
+  AUTOLOAD:=$(call AutoLoad,67,sun8i-codec-analog)
+  DEPENDS:=@TARGET_sunxi +kmod-sound-soc-core
+  $(call AddDepends/sound)
+endef
+
+define KernelPackage/sound-soc-sun8i-codec-analog/description
+  Kernel support for AllWinner built-in SoC audio sun8i-codec-analog
+endef
+
+$(eval $(call KernelPackage,sound-soc-sun8i-codec-analog))
--- a/package/boot/uboot-sunxi/Makefile	2019-01-30 14:34:49.000000000 +0200
+++ b/package/boot/uboot-sunxi/Makefile	2019-04-20 16:15:23.888581851 +0300
@@ -150,6 +150,12 @@
   BUILD_DEVICES:=sun8i-h2-plus-orangepi-r1
 endef
 
+define U-Boot/orangepi_zero
+  BUILD_SUBTARGET:=cortexa7
+  NAME:=Orange Pi Zero (H2+)
+  BUILD_DEVICES:=sun8i-h2-plus-orangepi-zero
+endef
+
 define U-Boot/orangepi_pc
   BUILD_SUBTARGET:=cortexa7
   NAME:=Orange Pi PC (H3)
@@ -230,6 +236,7 @@
 	nanopi_neo \
 	nanopi_neo_plus2 \
 	orangepi_r1 \
+	orangepi_zero \
 	orangepi_pc \
 	orangepi_plus \
 	orangepi_2 \
