--- a/package/boot/uboot-sunxi/Makefile	2019-01-30 14:34:49.000000000 +0200
+++ b/package/boot/uboot-sunxi/Makefile	2019-03-31 02:43:39.778188086 +0200
@@ -150,6 +150,12 @@
   BUILD_DEVICES:=sun8i-h2-plus-orangepi-r1
 endef
 
+define U-Boot/orangepi_zero
+  BUILD_SUBTARGET:=cortexa7
+  NAME:=Orange Pi Zero (H2+)
+  BUILD_DEVICES:=sun8i-h2-plus-orangepi-zero
+endef
+
 define U-Boot/orangepi_pc
   BUILD_SUBTARGET:=cortexa7
   NAME:=Orange Pi PC (H3)
@@ -230,6 +236,7 @@
 	nanopi_neo \
 	nanopi_neo_plus2 \
 	orangepi_r1 \
+	orangepi_zero \
 	orangepi_pc \
 	orangepi_plus \
 	orangepi_2 \
--- a/target/linux/sunxi/patches-4.14/061-dts-add-usb2-usb3-ir_pins-opi_zero.patch	1970-01-01 03:00:00.000000000 +0300
+++ b/target/linux/sunxi/patches-4.14/061-dts-add-usb2-usb3-ir_pins-opi_zero.patch	2019-04-12 14:15:32.924140529 +0300
@@ -0,0 +1,67 @@
+--- a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
++++ b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
+@@ -103,6 +103,14 @@
+ &ehci1 {
+ 	status = "okay";
+ };
++
++&ehci2 {
++	status = "okay";
++};
++
++&ehci3 {
++	status = "okay";
++};
+ 
+ &emac {
+ 	phy-handle = <&int_mii_phy>;
+@@ -150,17 +150,48 @@
+ &ohci1 {
+ 	status = "okay";
+ };
++
++&ohci2 {
++	status = "okay";
++};
++
++&ohci3 {
++	status = "okay";
++};
++
++&ir {
++	pinctrl-names = "default";
++	pinctrl-0 = <&ir_pins_a>;
++	status = "okay";
++};
+ 
+ &spi0 {
+ 	/* Disable SPI NOR by default: it optional on Orange Pi Zero boards */
+-	status = "disabled";
++	status = "okay";
+ 
+ 	flash@0 {
+ 		#address-cells = <1>;
+ 		#size-cells = <1>;
+ 		compatible = "mxicy,mx25l1606e", "winbond,w25q128";
+ 		reg = <0>;
+ 		spi-max-frequency = <40000000>;
++//		partitions {
++//			compatible = "fixed-partitions";
++//			#address-cells = <1>;
++//			#size-cells = <1>;
++//			partition@0 {
++//				label = "uboot";
++//				reg = <0x0 0x100000>;
++//			};
++//			partition@100000 {
++//				label = "env";
++//				reg = <0x100000 0x100000>;
++//			};
++//			partition@200000 {
++//				label = "data";
++//				reg = <0x200000 0x200000>;
++//			};
++//		};
+ 	};
+ };
+ 
--- a/target/linux/sunxi/patches-4.14/062-dts-wifi-add-xradio.patch.patch	1970-01-01 03:00:00.000000000 +0300
+++ b/target/linux/sunxi/patches-4.14/062-dts-wifi-add-xradio.patch.patch	2019-04-12 14:17:28.860715428 +0300
@@ -0,0 +1,47 @@
+diff --git a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
+index 454408b..df356b5 100644
+--- a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
++++ b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
+@@ -49,6 +49,7 @@
+ 
+ #include <dt-bindings/gpio/gpio.h>
+ #include <dt-bindings/input/input.h>
++#include <dt-bindings/pinctrl/sun4i-a10.h>
+ 
+ / {
+ 	model = "Xunlong Orange Pi Zero";
+@@ -88,6 +88,23 @@
+ 		enable-active-high;
+ 		gpio = <&pio 0 20 GPIO_ACTIVE_HIGH>;
+ 	};
++
++	reg_vdd_cpux: vdd-cpux-regulator {
++		compatible = "regulator-gpio";
++		regulator-name = "vdd-cpux";
++		regulator-type = "voltage";
++		regulator-boot-on;
++		regulator-always-on;
++		regulator-min-microvolt = <1100000>;
++		regulator-max-microvolt = <1300000>;
++		regulator-ramp-delay = <50>; /* 4ms */
++
++		gpios = <&r_pio 0 6 GPIO_ACTIVE_HIGH>; /* PL6 */
++		enable-active-high;
++		gpios-states = <1>;
++		states = <1100000 0
++			  1300000 1>;
++	};
+ 
+ 	wifi_pwrseq: wifi_pwrseq {
+ 		compatible = "mmc-pwrseq-simple";
+@@ -144,6 +144,10 @@
+ 	 */
+ 	xr819: sdio_wifi@1 {
+ 		reg = <1>;
++		compatible = "xradio,xr819";
++		interrupt-parent = <&pio>;
++		interrupts = <6 10 IRQ_TYPE_EDGE_RISING>;
++		interrupt-names = "host-wake";
+ 	};
+ };
+ 
--- a/target/linux/sunxi/image/cortex-a7.mk	2019-01-30 14:34:49.000000000 +0200
+++ b/target/linux/sunxi/image/cortex-a7.mk	2019-04-13 05:37:56.660274995 +0300
@@ -108,6 +108,16 @@
 TARGET_DEVICES += sun6i-a31-m9
 
 
+define Device/sun8i-h2-plus-orangepi-zero
+  DEVICE_TITLE:=Xunlong Orange Pi Zero
+  DEVICE_PACKAGES:=kmod-rtc-sunxi kmod-xradio
+  SUPPORTED_DEVICES:=xunlong,orangepi-zero
+  SUNXI_DTS:=sun8i-h2-plus-orangepi-zero
+endef
+
+TARGET_DEVICES += sun8i-h2-plus-orangepi-zero
+
+
 define Device/sun8i-h2-plus-orangepi-r1
   DEVICE_TITLE:=Xunlong Orange Pi R1
   DEVICE_PACKAGES:=kmod-rtc-sunxi kmod-usb-net kmod-usb-net-rtl8152
--- a/package/kernel/mac80211/Makefile	2019-01-30 14:34:49.000000000 +0200
+++ b/package/kernel/mac80211/Makefile	2019-04-13 19:35:32.192596287 +0300
@@ -10,7 +10,8 @@
 
 PKG_NAME:=mac80211
 
-PKG_VERSION:=2017-11-01
+#PKG_VERSION:=2017-11-01
+PKG_VERSION:=2019-03-13
 PKG_RELEASE:=9
 PKG_SOURCE_URL:=http://mirror2.openwrt.org/sources
 PKG_HASH:=8437ab7886b988c8152e7a4db30b7f41009e49a3b2cb863edd05da1ecd7eb05a
@@ -46,6 +47,7 @@
 	rtl8192ce rtl8192se rtl8192de rtl8192cu rtl8821ae \
 	rtl8xxxu \
 	wlcore wl12xx wl18xx \
+	xradio \
 	zd1211rw
 
 PKG_CONFIG_DEPENDS:= \
@@ -1563,6 +1565,13 @@
   AUTOLOAD:=$(call AutoProbe,zd1211rw)
 endef
 
+define KernelPackage/xradio
+  $(call KernelPackage/mac80211/Default)
+  TITLE:=Xradio xr819 test wifi-drivers support
+  DEPENDS+=+kmod-mac80211
+  FILES:=$(PKG_BUILD_DIR)/drivers/net/wireless/xradio/xradio_wlan.ko
+  AUTOLOAD:=$(call AutoProbe,xradio_wlan)
+endef
 
 
 config_package=$(if $(CONFIG_PACKAGE_kmod-$(1)),m)
@@ -1590,6 +1599,7 @@
 	WLAN_VENDOR_RSI \
 	WLAN_VENDOR_ST \
 	WLAN_VENDOR_TI \
+	WLAN_VENDOR_XRADIO \
 	WLAN_VENDOR_ZYDAS \
 
 config-$(call config_package,cfg80211) += CFG80211
@@ -1736,6 +1746,7 @@
 config-$(call config_package,wl12xx) += WL12XX
 config-$(call config_package,wl18xx) += WL18XX
 config-y += WL_TI WILINK_PLATFORM_DATA
+config-$(call config_package,xradio) += XRADIO
 config-$(call config_package,zd1211rw) += ZD1211RW
 config-$(call config_package,rsi91x) += RSI_91X
 config-$(call config_package,rsi91x-usb) += RSI_USB
@@ -1893,6 +1904,17 @@
 	$(INSTALL_DATA) $(PKG_BUILD_DIR)/$(ZD1211FW_NAME)/zd1211* $(1)/lib/firmware/zd1211
 endef
 
+define KernelPackage/xradio/install
+	$(INSTALL_DIR) $(1)/lib/modules/$(LINUX_VERSION)
+	$(INSTALL_BIN) \
+		$(PKG_BUILD_DIR)/drivers/net/wireless/xradio/xradio_wlan.ko \
+			$(1)/lib/modules/$(LINUX_VERSION)/
+
+	$(INSTALL_DIR) $(1)/lib/firmware/xr819
+	$(CP) $(PKG_BUILD_DIR)/drivers/net/wireless/xradio/xr819/*_xr819.bin \
+		$(1)/lib/firmware/xr819/
+endef
+
 
 $(eval $(call KernelPackage,adm8211))
 $(eval $(call KernelPackage,airo))
@@ -1970,4 +1992,5 @@
 $(eval $(call KernelPackage,wlcore))
 $(eval $(call KernelPackage,wl12xx))
 $(eval $(call KernelPackage,wl18xx))
+$(eval $(call KernelPackage,xradio))
 $(eval $(call KernelPackage,zd1211rw))
--- a/package/kernel/mac80211/files/lib/wifi/mac80211.sh	2019-01-30 14:34:49.000000000 +0200
+++ b/package/kernel/mac80211/files/lib/wifi/mac80211.sh	2019-02-18 00:41:10.256441878 +0200
@@ -1,6 +1,9 @@
 #!/bin/sh
 append DRIVERS "mac80211"
 
+board=$(cat /tmp/sysinfo/board_name)
+boardname="${board##*,}"
+
 lookup_phy() {
 	[ -n "$phy" ] && {
 		[ -d /sys/class/ieee80211/$phy ] && return
@@ -113,13 +116,13 @@
 			set wireless.radio${devidx}.hwmode=11${mode_band}
 			${dev_id}
 			${ht_capab}
-			set wireless.radio${devidx}.disabled=1
+			set wireless.radio${devidx}.disabled=
 
 			set wireless.default_radio${devidx}=wifi-iface
 			set wireless.default_radio${devidx}.device=radio${devidx}
 			set wireless.default_radio${devidx}.network=lan
 			set wireless.default_radio${devidx}.mode=ap
-			set wireless.default_radio${devidx}.ssid=OpenWrt
+			set wireless.default_radio${devidx}.ssid=${boardname}
 			set wireless.default_radio${devidx}.encryption=none
 EOF
 		uci -q commit wireless
