--- a/package/kernel/mac80211/Makefile	2019-01-30 14:34:49.000000000 +0200
+++ b/package/kernel/mac80211/Makefile	2019-04-13 19:35:32.192596287 +0300
@@ -10,7 +10,8 @@
 
 PKG_NAME:=mac80211
 
-PKG_VERSION:=2017-11-01
+#PKG_VERSION:=2017-11-01
+PKG_VERSION:=2019-03-13
 PKG_RELEASE:=9
 PKG_SOURCE_URL:=http://mirror2.openwrt.org/sources
 PKG_HASH:=8437ab7886b988c8152e7a4db30b7f41009e49a3b2cb863edd05da1ecd7eb05a
@@ -46,6 +47,7 @@
 	rtl8192ce rtl8192se rtl8192de rtl8192cu rtl8821ae \
 	rtl8xxxu \
 	wlcore wl12xx wl18xx \
+	xradio \
 	zd1211rw
 
 PKG_CONFIG_DEPENDS:= \
@@ -1563,6 +1565,13 @@
   AUTOLOAD:=$(call AutoProbe,zd1211rw)
 endef
 
+define KernelPackage/xradio
+  $(call KernelPackage/mac80211/Default)
+  TITLE:=Xradio xr819 test wifi-drivers support
+  DEPENDS+=+kmod-mac80211
+  FILES:=$(PKG_BUILD_DIR)/drivers/net/wireless/xradio/xradio_wlan.ko
+  AUTOLOAD:=$(call AutoProbe,xradio_wlan)
+endef
 
 
 config_package=$(if $(CONFIG_PACKAGE_kmod-$(1)),m)
@@ -1590,6 +1599,7 @@
 	WLAN_VENDOR_RSI \
 	WLAN_VENDOR_ST \
 	WLAN_VENDOR_TI \
+	WLAN_VENDOR_XRADIO \
 	WLAN_VENDOR_ZYDAS \
 
 config-$(call config_package,cfg80211) += CFG80211
@@ -1736,6 +1746,7 @@
 config-$(call config_package,wl12xx) += WL12XX
 config-$(call config_package,wl18xx) += WL18XX
 config-y += WL_TI WILINK_PLATFORM_DATA
+config-$(call config_package,xradio) += XRADIO
 config-$(call config_package,zd1211rw) += ZD1211RW
 config-$(call config_package,rsi91x) += RSI_91X
 config-$(call config_package,rsi91x-usb) += RSI_USB
@@ -1893,6 +1904,17 @@
 	$(INSTALL_DATA) $(PKG_BUILD_DIR)/$(ZD1211FW_NAME)/zd1211* $(1)/lib/firmware/zd1211
 endef
 
+define KernelPackage/xradio/install
+	$(INSTALL_DIR) $(1)/lib/modules/$(LINUX_VERSION)
+	$(INSTALL_BIN) \
+		$(PKG_BUILD_DIR)/drivers/net/wireless/xradio/xradio_wlan.ko \
+			$(1)/lib/modules/$(LINUX_VERSION)/
+
+	$(INSTALL_DIR) $(1)/lib/firmware/xr819
+	$(CP) $(PKG_BUILD_DIR)/drivers/net/wireless/xradio/xr819/*_xr819.bin \
+		$(1)/lib/firmware/xr819/
+endef
+
 
 $(eval $(call KernelPackage,adm8211))
 $(eval $(call KernelPackage,airo))
@@ -1970,4 +1992,5 @@
 $(eval $(call KernelPackage,wlcore))
 $(eval $(call KernelPackage,wl12xx))
 $(eval $(call KernelPackage,wl18xx))
+$(eval $(call KernelPackage,xradio))
 $(eval $(call KernelPackage,zd1211rw))
